File: content/exercises/intro-python/basics/hello-python/checker.py
import io
import subprocess
import sys
from pathlib import Path


def usage() -> None:
    raise SystemExit("Usage: checker.py <submission_path> <output_path>")


def main() -> None:
    if len(sys.argv) < 3:
        usage()

    submission_path = Path(sys.argv[1])
    output_path = Path(sys.argv[2])

    test_results: list[tuple[str, bool, str]] = []
    feedback: list[str] = []
    status = "failed"

    if not submission_path.exists():
        test_results.append(("submission exists", False, "Submission file was not found."))
        feedback.append("Submission file was not found.")
    else:
        test_results.append(("submission exists", True, "Submission file exists."))

        try:
            process = subprocess.run(
                [sys.executable, str(submission_path)],
                capture_output=True,
                text=True,
                check=False,
            )
        except OSError as error:
            test_results.append(("code runs", False, "Python could not execute the submission."))
            feedback.append(f"Execution error: {error}")
        else:
            if process.returncode != 0:
                test_results.append(("code runs", False, "Your code did not run successfully."))
                stderr_text = process.stderr.strip()
                feedback.append(f"Execution error: {stderr_text or 'unknown runtime error'}")
            else:
                test_results.append(("code runs", True, "Code executed without errors."))
                output_lines = [line.strip() for line in process.stdout.splitlines() if line.strip()]
                expected_line = "Hello, Python!"
                if expected_line in output_lines:
                    test_results.append(
                        ("prints expected greeting", True, "Output includes 'Hello, Python!'.")
                    )
                    status = "passed"
                    feedback.append("Correct solution submitted.")
                else:
                    test_results.append(
                        (
                            "prints expected greeting",
                            False,
                            "Use print() to output exactly 'Hello, Python!'.",
                        )
                    )
                    feedback.append("Use print() to output exactly 'Hello, Python!'.")

    output_stream = io.StringIO()
    output_stream.write(f"STATUS:{status}\n")
    for item in feedback:
        output_stream.write(f"FEEDBACK:{item}\n")
    for name, passed, message in test_results:
        output_stream.write(f"TEST:{name}|{'pass' if passed else 'fail'}|{message}\n")

    output_path.write_text(output_stream.getvalue(), encoding="utf-8")


if __name__ == "__main__":
    main()

File: content/exercises/intermediate-r/ggplot2/bar-chart/checker.R
args <- commandArgs(trailingOnly = TRUE)
if (length(args) < 2) stop("Usage: checker.R <submission_path> <output_path>")
submission_path <- args[[1]]
output_path <- args[[2]]

test_results <- list()
feedback <- c()
status <- "failed"
record_test <- function(name, passed, message) {
  test_results[[length(test_results) + 1]] <<- list(name = name, passed = passed, message = message)
}

execution_ok <- TRUE
execution_error <- NULL
last_value <- NULL
tryCatch({
  sourced <- source(submission_path, local = new.env(parent = baseenv()))
  last_value <- sourced$value
}, error = function(e) {
  execution_ok <<- FALSE
  execution_error <<- conditionMessage(e)
})

if (!execution_ok) {
  record_test("code runs", FALSE, "Your code did not run successfully.")
  feedback <- c(feedback, paste("Execution error:", execution_error))
} else {
  record_test("code runs", TRUE, "Code executed without errors.")
  is_plot <- inherits(last_value, "ggplot")
  record_test("returns ggplot", is_plot, "Return a ggplot bar chart object.")
}

passed_all <- length(test_results) > 0 && all(vapply(test_results, function(t) isTRUE(t$passed), logical(1)))
if (passed_all) {
  status <- "passed"; feedback <- c(feedback, "Correct solution submitted.")
} else if (length(feedback) == 0) {
  feedback <- c(feedback, "Create grouped summary and return ggplot bar chart with reference line.")
}

output_lines <- c(paste0("STATUS:", status))
for (msg in feedback) output_lines <- c(output_lines, paste0("FEEDBACK:", msg))
for (t in test_results) output_lines <- c(output_lines, paste0("TEST:", t$name, "|", if (isTRUE(t$passed)) "pass" else "fail", "|", t$message))
writeLines(output_lines, con = output_path)
